datafusion/datasource-parquet/src/opener.rs

```diff
@@ -524,6 +539,84 @@ fn should_enable_page_index(
             .unwrap_or(false)
 }
 
+use datafusion_physical_expr::expressions;
+
+/// Given a [`PhysicalExpr`] and a [`SchemaRef`], returns a new [`PhysicalExpr`] that
+/// is cast to the specified data type.
+/// Preference is always given to casting literal values to the data type of the column
+/// since casting the column to the literal value's data type can be significantly more expensive.
+/// Given two columns the cast is applied arbitrarily to the first column.
+pub fn cast_expr_to_schema(
+    expr: Arc<dyn PhysicalExpr>,
+    physical_file_schema: &Schema,
+    logical_file_schema: &Schema,
+    partition_values: Vec<ScalarValue>,
+    partition_fields: &[FieldRef],
+) -> Result<Arc<dyn PhysicalExpr>> {
+    expr.transform(|expr| {
+        if let Some(column) = expr.as_any().downcast_ref::<expressions::Column>() {
+            let logical_field = match logical_file_schema.field_with_name(column.name()) {
```

Comment:
nit: we could reduce a level of nesting with something like
```suggestion
        let Some(column) = expr.as_any().downcast_ref::<expressions::Column>() else {
            return Ok(Transformed::no(expr))
         }
```